firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java

```diff
@@ -45,8 +45,12 @@ public DefaultGrpcMetadata(
   @Override
   public void updateMetadata(@NonNull Metadata metadata) {
     metadata.put(USER_AGENT_HEADER, userAgentPublisher.getUserAgent());
-    metadata.put(
-        HEART_BEAT_HEADER,
-        Integer.toString(heartBeatInfo.getHeartBeatCode(firebaseFirestoreHeartBeatTag).getCode()));
+    int heartBeatCode = heartBeatInfo.getHeartBeatCode(firebaseFirestoreHeartBeatTag).getCode();
```

Comment:
How costly is calling `getHeartBeatCode`? Does it perform any I/O? Does it talk to the network?

Reply:
it looks into the file stored in the app to check when the last time the heartbeat was sent. It doesnt talk to network.

Reply:
it uses `SharedPreferences` so technically it only reads from disk once and keeps/manipulates state in memory, syncing to disk asynchronously.